dsync and dcp: allow user to control syncing of xattrs

ofaaland opened the pull request:
Rather than always copying xattrs, introduce a new option
"--xattrs | -e" to allow the user to control how this is done.

Options are
  "none"        Copy no xattrs
  "all"         Copy all xattrs (prior default)
  "non-lustre"  Copy non-lustre xattrs (new default)
  "libattr"        Copy xattrs not indicated as "skip" in /etc/xattr.conf

Copying xattrs is now independent of --preserve for owner, perms,
etc.

Lustre uses xattrs to record important information such as how the file
data should be distributed among the Lustre servers.  As a result,
copying the lustre xattrs prevents certain default values from taking
effect.

If "non-lustre" option is in effect, xattrs named in lustre
source file lustre_idl.h as of this writing are not synced (see
XATTR_NAME_SOM and friends).  In addition, xattrs with prefix
"lustre" are not synced.

Signed-off-by: Olaf Faaland <faaland1@llnl.gov>

adammoody commented:
@ofaaland , there are too many issues to track.  No problem that you weren't aware.  I just wanted to list them here because your work here will help address them.

We can also add these separately in dcp if you want.

We still have dcp1 and dcp because they implement two different algorithms.  Depending on the situation, one can be faster than the other.  Eventually, it'd be nice to roll both algorithms into one tool, but that's still future work.

adilger commented on src/dcp/dcp.c:
```diff
@@ -150,6 +151,7 @@ int main(int argc, char** argv)
         {"daos-preserve"        , required_argument, 0, 'D'},
         {"input"                , required_argument, 0, 'i'},
         {"chunksize"            , required_argument, 0, 'k'},
+        {"copy-xattrs"          , required_argument, 0, 'e'},
```
minor nit - both tar and rsync use "--xattrs" (also short option '-X' for rsync) for this option.  Any reason not to be consistent?

ofaaland replied:
Yes, --xattrs / -X would be better

adilger commented on src/dcp/dcp.c:
```diff
@@ -231,6 +234,23 @@ int main(int argc, char** argv)
                 break;
+            case 'e':
+                copy_xattrs_used = 1;
+                if (!strcmp(optarg,"none")) {
```
Not sure what the coding guidelines are for MPIFileutils, but (IMHO) the usage of "!strcmp()" can be misread, as it looks like it is checking "if not a good string compare".  For Lustre we prefer to use "strcmp() == 0", but by no means am I trying to push my code style upon you, just a thought.

adilger commented on src/dcp/dcp.c:
```diff
@@ -231,6 +234,23 @@ int main(int argc, char** argv)
+                    mfu_copy_opts->copy_xattrs = XATTR_COPY_ALL;
+                } else {
+                    if (rank == 0) {
+                        MFU_LOG(MFU_LOG_ERR, "Unrecognized option for --copy-xattrs");
```
It error message would be more useful if it printed the actual bad option inside single quotes, in case it is accidentally a filename or similar.

adilger commented on src/dcp/dcp.c:
```diff
@@ -332,6 +352,14 @@ int main(int argc, char** argv)
+    /* copy_xattrs only vaid with preserve */
+    if (mfu_copy_opts->preserve == false && copy_xattrs_used == 1) {
+        if (rank == 0) {
+            MFU_LOG(MFU_LOG_ERR, "Copy-xattrs may only be used with --preserve");
```
IMHO, errors like this are counter productive for users.  I just means if the user _wants_ to copy xattrs, they will hit an error, and then have to fix what the program could have done itself - enable the preserve option if --xattrs is used.

adilger commented on src/dcp/dcp.c:
```diff
@@ -231,6 +234,23 @@ int main(int argc, char** argv)
+                copy_xattrs_used = 1;
+                if (!strcmp(optarg,"none")) {
+                    mfu_copy_opts->copy_xattrs = XATTR_COPY_NONE;
+                } else if (!strcmp(optarg,"non-lustre")) {
```
For code like this, I've found it more useful to have an array of { value, "string" } for all of the options (with an empty entry at the end), and then loop over the array looking for the matching string to get the value (possibly in a small helper function), or in the case of the usage message loop over the array and print all of the strings.  That avoids duplicating the string/value mapping a few times (as is done here and dsync.c), and avoids future errors if a new type is added but the usage message is not updated, or not updated in both programs, or similar.

adilger reviewed (commented):
None of my comments are particularly related to the functionality being added here (which looks fine to me, the patch seems rather straight forward).  They are mostly generic comments about code style, and trying to minimize/avoid potential future bugs related to the changes being made here.

adammoody commented on man/dcp.1:
```diff
@@ -60,6 +60,17 @@ The default chunksize is 64MB.
 .UNINDENT
 .INDENT 0.0
 .TP
+.B \-\-xattrs WHICH
+Copy extended attributes ("xattrs") from source files to target files.  WHICH
+determines which xattrs are copied.  Options are to copy no xattrs, all xattrs,
+xattrs not excluded by /etc/xattr.conf, or all xattrs except those which have
+special meaning to lustre.  Certain xattrs control lustre features on a
+file-by-file basis, such as how the file data is distributed across Lustre
+servers.  Values must be in {none, all, libattr, non-lustre}.  The default is
+non-lustre.
+.UNINDENT
+.INDENT 0.0
+.TP
```
Would you please move this to ``doc/rst/dcp1.rst``.  We auto-generate these man pages from the rst source files.

Let's also capitalize "Lustre" in the user-facing text.

adammoody replied:
I can help generate the man pages from there.

ofaaland replied:
Cool, I didn't notice the rst files.  So much easier to read and edit than nroff or whatever that is.

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -330,18 +332,44 @@ static int mfu_copy_xattrs(
+            } else if (copy_opts->copy_xattrs == XATTR_COPY_NONE) {
+                copy_xattr = 0;
+            } else if (copy_opts->copy_xattrs == XATTR_USE_LIBATTR) {
+                if (attr_copy_action(name, &ctx) == ATTR_ACTION_SKIP)
+                    copy_xattr = 0;
```
Coding style: we try to always use braces, even for one-liners.
```suggestion
                if (attr_copy_action(name, &ctx) == ATTR_ACTION_SKIP) {
                    copy_xattr = 0;
                }
```

ofaaland replied:
Done

adammoody commented on src/common/mfu_param_path.c:
```diff
@@ -303,6 +303,27 @@ static void mfu_unpack_param(const char** pptr, mfu_param_path* param)
+ */
+attr_copy_t parse_copy_xattrs_option(char *optarg)
+{
+    if (strcmp(optarg,"none") == 0)
+        return XATTR_COPY_NONE;
+
+    if (strcmp(optarg,"non-lustre") == 0)
+        return XATTR_SKIP_LUSTRE;
+
+    if (strcmp(optarg,"libattr") == 0)
+        return XATTR_USE_LIBATTR;
+
+    if (strcmp(optarg,"all") == 0)
+        return XATTR_COPY_ALL;
```
style: let's include braces through here, too

ofaaland replied:
done

adammoody commented on src/dcp/dcp.c:
```diff
@@ -79,6 +79,7 @@ void print_usage(void)
 #endif
     printf("  -b, --bufsize <SIZE>     - IO buffer size in bytes (default " MFU_BUFFER_SIZE_STR ")\n");
     printf("  -k, --chunksize <SIZE>   - work size per task in bytes (default " MFU_CHUNK_SIZE_STR ")\n");
+    printf("  -e, --xattrs <OPT>       - copy xattrs (none, all, non-lustre, libattr)\n");
```
Do you mean to use ``-X`` now?
```suggestion
    printf("  -X, --xattrs <OPT>       - copy xattrs (none, all, non-lustre, libattr)\n");
```

ofaaland replied:
Yes

adammoody commented on src/dcp/dcp.c:
```diff
@@ -90,7 +91,7 @@ void print_usage(void)
     printf("  -L, --dereference        - copy original files instead of links\n");
     printf("  -P, --no-dereference     - don't follow links in source\n");
-    printf("  -p, --preserve           - preserve permissions, ownership, timestamps, extended attributes\n");
+    printf("  -p, --preserve           - preserve permissions, ownership, timestamps (see also --copy-xattrs)\n");
```
Do you mean this instead?
```suggestion
    printf("  -p, --preserve           - preserve permissions, ownership, timestamps (see also --xattrs)\n");
```

ofaaland replied:
Yep

adammoody commented on src/dsync/dsync.c:
```diff
@@ -3047,7 +3049,7 @@ int main(int argc, char **argv)
     while (1) {
         int c = getopt_long(
-            argc, argv, "b:cDso:LPSvqh",
+            argc, argv, "b:cDe:so:LPSvqh",
```
Do you mean to have ``X:`` in here instead of ``e:``?

ofaaland replied:
Yes

adammoody commented:
Thanks again @ofaaland .  Sorry to be so tardy on this.  Great work.  I found a few things to check, but otherwise it looks good.

We use sphinx to generate our man pages from the rst files.  I can help you set that up.

ofaaland commented:
Hi @adammoody , I fixed up the issues you found, and also added a test.  I ran the test bash file by hand, rather than using nose2 and the python test-runner because I'm not sure how it works - but it should work, I've included a python wrapper copied from the dcp test.

I found doc/README.md and followed the instructions to setup and run Sphinx.  It ran, and it updated the man pages to reflect my .rst updates, but it also made a bunch of markup changes that are unrelated to my work.  So perhaps the version of Sphinx is different or there's some other Sphinx configuration I haven't done.  Here's an example:
```
 .TP
 .B \-b, \-\-blocksize SIZE
 Set the compression block size, from 1 to 9.
-Where 1=100kB <E2><80><A6> and 9=900kB. Default is 9.
+Where 1=100kB ... and 9=900kB. Default is 9.
 .UNINDENT
 .INDENT 0.0
 .TP
```

Because I don't know the reason for those markup changes, I've discarded the changes Sphinx made for now.  Let me know how to proceed.

Thanks!

ofaaland commented:
dsync test output:
```
Testing dsync
PASSED verify of option none for /mnt/xfs/dest type xfs
PASSED verify of option all for /mnt/xfs/dest type xfs
SKIPPED verify of option non-lustre
PASSED verify of option libattr for /mnt/xfs/dest type xfs
```

adammoody commented:
Thanks @ofaaland .  That's great!

I just gave you a couple of files that I used to install sphinx and then build the man pages.  It's been a while since I tried myself, so it could just be due to a newer sphinx version like you say.

We could also merge without the changes to the man files and generate the new man pages in a later commit if you want.  I often only update the rst files in my commits and forget all about needing to regenerate the man pages until we stamp a new release.

adammoody commented on src/dsync/dsync.c:
```diff
@@ -63,6 +63,7 @@ static void print_usage(void)
     printf("  -b  --batch-files <N>   - batch files into groups of N during copy\n");
     printf("      --bufsize <SIZE>    - IO buffer size in bytes (default " MFU_BUFFER_SIZE_STR ")\n");
     printf("      --chunksize <SIZE>  - minimum work size per task in bytes (default " MFU_CHUNK_SIZE_STR ")\n");
+    printf("  -e, --xattrs <OPT>      - copy xattrs (none, all, non-lustre, libattr)\n");
```
We've got ``e`` instead of ``X`` here, too

ofaaland replied:
oops, fixed.

adammoody commented on src/dcp/dcp.c:
```diff
@@ -168,7 +170,7 @@ int main(int argc, char** argv)
     while(1) {
         int c = getopt_long(
-                    argc, argv, "b:d:g:i:k:LPpsSvqh",
+                    argc, argv, "b:d:e:g:i:k:LPpsSvqh",
```
And an ``X:`` here?

ofaaland replied:
Yep

ofaaland commented:
OK, I fixed up those issues, removed a couple of whitespace issues that snuck in, and hand-tested dcp to make sure -X works.  I think it's ready to go.

I think it makes sense to update the man pages in a separate commit, in case there are other unrelated man page changes.  But I'll take your sphinx files and see if I still get markup changes.

Thanks @adammoody

ofaaland commented:
Looks like the markup changes are not due to the process I followed - I get the same results using your scripts.  There is at least one other content change (for DAOS) to the .rst files that isn't yet reflected in the man pages, so I think a man-page-only commit makes sense.

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -993,7 +1024,7 @@ static int mfu_create_directory(
 
     /* copy extended attributes on directory */
-    if (copy_opts->preserve) {
+    if (copy_opts->copy_xattrs) {
```
It seems like this condition might always be true, since I think ``XATTR_COPY_NONE=1`` in the enum.  I see the function handles NONE correctly, though it might best to avoid calling all together when disabled.  Maybe replace this with ``if (copy_opts->copy_xattrs != XATTR_COPY_NONE)``?

ofaaland replied:
Yes, good catch.  I forgot to change it after I switched copy_xattrs from a boolean to an enum, but the mistake was hidden because the function handled NONE correctly.

I changed that condition there and in the other two places mfu_copy_xattrs() is called as you suggested.  I also simplified the logic within mfu_copy_xattrs(), and then tested to confirm it still worked.

adammoody commented:
Thanks, @ofaaland .  This is good to go, but we've got a conflict on some rst files due to a change that snuck in.  Would you please update to resolve those, and then we can merge?

ofaaland commented:
Thanks @adammoody.  Rebased on master and fixed the conflicts.

adammoody commented:
Really nice addition, @ofaaland !
